File: gateways/tunnel-agent/src/auth.test.ts
import { describe, it, expect } from 'vitest';
import { buildWsOptions } from './auth';
import type { TunnelConfig } from './config';

function makeConfig(overrides: Partial<TunnelConfig> = {}): TunnelConfig {
  return {
    serverUrl: 'wss://example.com/tunnel',
    token: 'test-token-abc',
    gatewayId: 'gw-001',
    agentVersion: '1.0.0',
    pingIntervalMs: 15000,
    reconnectInitialMs: 1000,
    reconnectMaxMs: 60000,
    localServiceHost: 'localhost',
    localServicePort: 4822,
    ...overrides,
  };
}

describe('buildWsOptions', () => {
  it('returns Authorization header with Bearer prefix', () => {
    const opts = buildWsOptions(makeConfig({ token: 'my-secret-token' }));
    const headers = opts.headers as Record<string, string>;
    expect(headers.Authorization).toBe('Bearer my-secret-token');
  });

  it('includes X-Gateway-Id header', () => {
    const opts = buildWsOptions(makeConfig({ gatewayId: 'gw-42' }));
    const headers = opts.headers as Record<string, string>;
    expect(headers['X-Gateway-Id']).toBe('gw-42');
  });

  it('includes X-Agent-Version header', () => {
    const opts = buildWsOptions(makeConfig({ agentVersion: '2.3.4' }));
    const headers = opts.headers as Record<string, string>;
    expect(headers['X-Agent-Version']).toBe('2.3.4');
  });

  it('sets handshakeTimeout to 10 seconds', () => {
    const opts = buildWsOptions(makeConfig());
    expect(opts.handshakeTimeout).toBe(10_000);
  });

  it('includes TLS certs when all are provided', () => {
    const opts = buildWsOptions(makeConfig({
      caCert: '-----BEGIN CERTIFICATE-----\nCA\n-----END CERTIFICATE-----',
      clientCert: '-----BEGIN CERTIFICATE-----\nCLIENT\n-----END CERTIFICATE-----',
      clientKey: '-----BEGIN PRIVATE KEY-----\nKEY\n-----END PRIVATE KEY-----',
    }));
    expect(opts.ca).toBe('-----BEGIN CERTIFICATE-----\nCA\n-----END CERTIFICATE-----');
    expect(opts.cert).toBe('-----BEGIN CERTIFICATE-----\nCLIENT\n-----END CERTIFICATE-----');
    expect(opts.key).toBe('-----BEGIN PRIVATE KEY-----\nKEY\n-----END PRIVATE KEY-----');
  });

  it('includes only CA cert when client cert/key are absent', () => {
    const opts = buildWsOptions(makeConfig({
      caCert: '-----BEGIN CERTIFICATE-----\nCA\n-----END CERTIFICATE-----',
    }));
    expect(opts.ca).toBe('-----BEGIN CERTIFICATE-----\nCA\n-----END CERTIFICATE-----');
    expect(opts.cert).toBeUndefined();
    expect(opts.key).toBeUndefined();
  });

  it('omits all TLS fields when no certs are provided', () => {
    const opts = buildWsOptions(makeConfig());
    expect(opts.ca).toBeUndefined();
    expect(opts.cert).toBeUndefined();
    expect(opts.key).toBeUndefined();
  });

  it('does not include client cert/key when only one of them is set', () => {
    const opts = buildWsOptions(makeConfig({
      clientCert: '-----BEGIN CERTIFICATE-----\nCLIENT\n-----END CERTIFICATE-----',
      // clientKey intentionally missing
    }));
    expect(opts.cert).toBeUndefined();
    expect(opts.key).toBeUndefined();
  });
});

File: gateways/tunnel-agent/auth_test.go
package main

import (
	"net/url"
	"strings"
	"testing"
)

func TestBuildAuthHeaders(t *testing.T) {
	cfg := testConfig()
	cfg.ClientCert = "-----BEGIN CERTIFICATE-----\nabc+123\n-----END CERTIFICATE-----"

	headers := buildAuthHeaders(cfg)
	if got := headers.Get("Authorization"); got != "Bearer token" {
		t.Fatalf("unexpected Authorization header %q", got)
	}
	if got := headers.Get("X-Gateway-Id"); got != "gw-1" {
		t.Fatalf("unexpected X-Gateway-Id header %q", got)
	}
	if got := headers.Get("X-Agent-Version"); got != defaultAgentVersion {
		t.Fatalf("unexpected X-Agent-Version header %q", got)
	}
	decoded, err := url.QueryUnescape(headers.Get("X-Client-Cert"))
	if err != nil {
		t.Fatalf("client cert header is not URL encoded: %v", err)
	}
	if decoded != cfg.ClientCert {
		t.Fatalf("unexpected decoded client cert header %q", decoded)
	}
}

func TestBuildAuthHeadersDoesNotPutTokenInURLFields(t *testing.T) {
	cfg := testConfig()
	cfg.Token = "super-secret-token"
	headers := buildAuthHeaders(cfg)
	if strings.Contains(cfg.ServerURL, cfg.Token) {
		t.Fatalf("test config unexpectedly contains token in URL")
	}
	if headers.Get("Authorization") != "Bearer super-secret-token" {
		t.Fatalf("token must only appear in Authorization header")
	}
}

func TestBuildTLSConfigOmitTLSWhenNoCertsProvided(t *testing.T) {
	tlsConfig, err := buildTLSConfig(testConfig())
	if err != nil {
		t.Fatalf("buildTLSConfig returned error: %v", err)
	}
	if tlsConfig != nil {
		t.Fatalf("expected nil TLS config")
	}
}

func TestBuildTLSConfigRejectsInvalidCA(t *testing.T) {
	cfg := testConfig()
	cfg.CACert = "not a certificate"
	if _, err := buildTLSConfig(cfg); err == nil {
		t.Fatalf("expected invalid CA error")
	}
}

func testConfig() *tunnelConfig {
	return &tunnelConfig{
		ServerURL:        "ws://example.test/api/tunnel/connect",
		Token:            "token",
		GatewayID:        "gw-1",
		AgentVersion:     defaultAgentVersion,
		PingInterval:     15_000_000_000,
		ReconnectInitial: 1_000_000_000,
		ReconnectMax:     60_000_000_000,
		LocalServiceHost: "127.0.0.1",
		LocalServicePort: 4822,
	}
}
